clash-cores/src/Clash/Cores/Xilinx/VIO/Internal/BlackBoxes.hs

```diff
@@ -0,0 +1,411 @@
+{-|
+  Copyright   :  (C) 2022 Google Inc
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+
+  Blackbox implementation for primitives in "Clash.Cores.Xilinx.VIO".
+-}
+
+{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Clash.Cores.Xilinx.VIO.Internal.BlackBoxes (vioProbeBBF) where
+
+import Prelude
+
+import GHC.Stack (HasCallStack)
+
+import Data.Maybe (isNothing)
+import Data.Foldable (fold)
+import Data.Tree (Tree(..))
+import Data.Bits (testBit, setBit, zeroBits)
+import Data.Text.Prettyprint.Doc.Extra (Doc)
+import qualified Data.Text as T (pack, unpack, append, concat)
+
+import Control.Arrow (first)
+import Control.Monad (when, forM)
+import Control.Monad.State (State)
+import Control.Exception (assert)
+
+import qualified Clash.Netlist.Id as Id
+import qualified Clash.Primitives.DSL as DSL
+import Clash.Backend (Backend)
+import Clash.Netlist.Types
+  ( Size
+  , Bit(..)
+  , Expr(..)
+  , HWType(..)
+  , Literal(..)
+  , Modifier(..)
+  , EntityOrComponent(..)
+  , TemplateFunction(..)
+  , BlackBoxContext(..)
+  , BlackBox(BBFunction)
+  )
+import Clash.Netlist.BlackBox.Types
+  ( TemplateKind(..)
+  , RenderVoid(..)
+  , BlackBoxFunction
+  , BlackBoxMeta(..)
+  , emptyBlackBoxMeta
+  )
+
+import Clash.Cores.Xilinx.Internal
+  ( TclPurpose(..)
+  , IpConfig(..)
+  , defIpConfig
+  , property
+  , renderTcl
+  )
+
+vioProbeBBF :: HasCallStack => BlackBoxFunction
+vioProbeBBF _isD _primName _args _resTys = pure $ Right (bbMeta, bb)
+ where
+   bbMeta :: BlackBoxMeta
+   bbMeta = emptyBlackBoxMeta
+     { bbKind = TDecl
+     , bbRenderVoid = RenderVoid
+     , bbIncludes =
+         [ ( ("vioProbe", "clash.tcl")
+           , BBFunction (show 'vioProbeTclTF) 0 vioProbeTclTF
+           )
+         ]
+     }
+
+   bb :: BlackBox
+   bb = BBFunction (show 'vioProbeTF) 0 vioProbeTF
+
+vioProbeTF :: HasCallStack => TemplateFunction
+vioProbeTF =
+  TemplateFunction
+    (
+--      0 :       -- KnownDomain dom
+--      1 :       -- VIO dom a o
+--      2 :       -- initialOutputProbeValues
+      3 :       -- Clock dom
+      [4,5..]   -- input probes
+    )
+    -- 'validateVioProbeBCC' already produces string describing
+    -- failing checks, but 'TemplateFunction' cannot handle this
+    -- yet. This is prepared to get updated easily as soon as the
+    -- feature gets implemented in clash-lib.
+    (isNothing . validateVioProbeBBC)
```

Comment:
There is a note somewhere in `clash-lib` that the validation function should be updated to produce an error string instead of only returning a `Bool`. This is why I already prepared the `validateVioProbeBBC` for that. Is there any guideline telling you when it is ok to use `error` or when I have to use the clash error reporting instead?

This implementation is not really consistent in terms of error handling in any sense right now. Hence, that's definitely an open todo for further improvements.

Reply:
> Is there any guideline telling you when it is ok to use error or when I have to use the clash error reporting instead?

Not really; error handling withing `clash-lib` is a mess. The most important thing is to get useful error messages to users though, which `error` allows you to do (despite being :nauseated_face:).